// houdini/build/runtime/stores/fragment.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.fragmentStore = void 0;
// externals
const store_1 = require("svelte/store");
const pagination_1 = require("../lib/pagination");
// a fragment store exists in multiple places in a given application so we
// can't just return a store directly, the user has to load the version of the
// fragment store for the object the store has been mixed into
function fragmentStore({ artifact, config, paginatedArtifact, paginationMethods, }) {
    return {
        name: artifact.name,
        get(initialValue) {
            // at the moment a fragment store doesn't really do anything
            // but we're going to keep it wrapped in a store so we can eventually
            // optimize the updates
            const fragmentStore = (0, store_1.writable)(initialValue);
            // build up the methods we want to use
            let extraMethods = {};
            if (paginatedArtifact) {
                const handlers = (0, pagination_1.fragmentHandlers)({
                    config,
                    paginationArtifact: paginatedArtifact,
                    initialValue,
                    store: fragmentStore,
                });
                extraMethods = Object.fromEntries(Object.entries(paginationMethods).map(([key, value]) => [key, handlers[value]]));
            }
            return {
                subscribe: fragmentStore.subscribe,
                update: fragmentStore.set,
                ...extraMethods,
            };
        },
    };
}
exports.fragmentStore = fragmentStore;
