// houdini/build/runtime/inline/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.subscription = exports.paginatedFragment = exports.fragment = exports.mutation = exports.paginatedQuery = exports.query = void 0;
var query_1 = require("./query");
Object.defineProperty(exports, "query", { enumerable: true, get: function () { return query_1.query; } });
Object.defineProperty(exports, "paginatedQuery", { enumerable: true, get: function () { return query_1.paginatedQuery; } });
var mutation_1 = require("./mutation");
Object.defineProperty(exports, "mutation", { enumerable: true, get: function () { return mutation_1.mutation; } });
var fragment_1 = require("./fragment");
Object.defineProperty(exports, "fragment", { enumerable: true, get: function () { return fragment_1.fragment; } });
Object.defineProperty(exports, "paginatedFragment", { enumerable: true, get: function () { return fragment_1.paginatedFragment; } });
var subscription_1 = require("./subscription");
Object.defineProperty(exports, "subscription", { enumerable: true, get: function () { return subscription_1.subscription; } });
