// houdini/build/cmd/transforms/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.fragmentVariables = exports.paginate = exports.addID = exports.typename = exports.list = exports.internalSchema = exports.composeQueries = void 0;
var composeQueries_1 = require("./composeQueries");
Object.defineProperty(exports, "composeQueries", { enumerable: true, get: function () { return __importDefault(composeQueries_1).default; } });
var schema_1 = require("./schema");
Object.defineProperty(exports, "internalSchema", { enumerable: true, get: function () { return __importDefault(schema_1).default; } });
var list_1 = require("./list");
Object.defineProperty(exports, "list", { enumerable: true, get: function () { return __importDefault(list_1).default; } });
var typename_1 = require("./typename");
Object.defineProperty(exports, "typename", { enumerable: true, get: function () { return __importDefault(typename_1).default; } });
var addID_1 = require("./addID");
Object.defineProperty(exports, "addID", { enumerable: true, get: function () { return __importDefault(addID_1).default; } });
var paginate_1 = require("./paginate");
Object.defineProperty(exports, "paginate", { enumerable: true, get: function () { return __importDefault(paginate_1).default; } });
var fragmentVariables_1 = require("./fragmentVariables");
Object.defineProperty(exports, "fragmentVariables", { enumerable: true, get: function () { return __importDefault(fragmentVariables_1).default; } });
